### PR TITLE
csclient: add UploadCharmWithRevision

### PR DESCRIPTION
**Pull request by rogpeppe**
I changed my mind about the previous PR, when I remembered about
promulgation. This reverts the original methods to their previous behaviour
and adds additional methods to upload with a specific revision and promulgated
revision.

### CODE REVIEW

**Comment by frankban** [comment]
:+1:

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/924/
Test FAILed.

File: csclient/csclient.go [hunk 1]
```diff
@@ -163,53 +163,89 @@ func (c *Client) GetArchive(id *charm.Reference) (r io.ReadCloser, eid *charm.Re
...
+	if err != nil {
+		return errgo.Notef(err, "cannot open charm archive")
+	}
+	defer r.Close()
+	_, err = c.uploadArchive(id, r, hash, size, promulgatedRevision)
+	return errgo.Mask(err)
 }
 
-// UploadBundle uploads the given bundle to the charm store with the given id.
+// UploadCharm uploads the given charm to the charm store with the given id,
```
Comment (mhilton) [inline csclient/csclient.go]: This should be UploadBundle

**Comment by mhilton** [comment]
:+1: LGTM with comment fixed

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/925/
Test FAILed.

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/charmstore-merge/302

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**